### PR TITLE
[Opt merge part2] introduce cascade merge algorithm

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Changes
Introduce a cascade merge algorithm:
1. Two-way stream merge: cutoff streaming chunk into small morsels, merge each morsel individually
2. Multi-way cascade merge: implement cascade merge based on two-way merge

### CODE REVIEW

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

File: be/src/exec/vectorized/sorting/merge.h [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    const Column* get_column(int index) const { return orderby[index].get(); }
+    bool empty() const { return range.second == range.first; }
+    void reset() {
+        chunk->reset();
+        orderby.clear();
+        range = {};
+    }
+    ChunkUniquePtr clone_chunk() {
+        if (range.first == 0) {
+            return chunk->clone_unique();
```
Comment (trueeyu) [inline be/src/exec/vectorized/sorting/merge.h:46]: why clone unique directly if `firs==0`

File: be/src/exec/vectorized/sorting/merge.h [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        range = {};
+    }
+
+    // Clone this SortedRun, could be the entire chunk or slice of chunk
+    ChunkUniquePtr clone_slice() {
+        if (range.first == 0 && range.second == chunk->num_rows()) {
+            return chunk->clone_unique();
+        } else {
+            size_t slice_rows = num_rows();
+            ChunkUniquePtr cloned = chunk->clone_empty(slice_rows - range.first);
```
Comment (trueeyu) [inline be/src/exec/vectorized/sorting/merge.h:51]: why `- rang.first`

Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/merge.h:51]: fixed.

**Review by trueeyu — DISMISSED** [review]

**Review by Pslydhh — DISMISSED** [review]

**Review by kangkaisen — DISMISSED** [review]

**Review by stdpain — DISMISSED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Pull request merged**